File: src/mk3/JavaPackage/src/main/java/tamaf/WakerBehavior.java
package tamaf;

public abstract class WakerBehavior extends Behavior {
    private long wakeUpTime;
    private boolean finished = false;

    public WakerBehavior(long timeoutMs) {
        super();
        this.wakeUpTime = System.currentTimeMillis() + timeoutMs;
    }

    public WakerBehavior(Agent agent, long timeoutMs) {
        super(agent);
        this.wakeUpTime = System.currentTimeMillis() + timeoutMs;
    }

    @Override
    public void action() {
        if (System.currentTimeMillis() >= wakeUpTime) {
            onWake();
            finished = true;
        }
    }

    public abstract void onWake();

    public long getRemainingTime() {
        return Math.max(0, wakeUpTime - System.currentTimeMillis());
    }

    @Override
    public boolean done() {
        return finished;
    }
}

File: src/mk1/esp32package/src/tamaf/behaviors/WakerBehavior.cpp
#include "WakerBehavior.h"

namespace tamaf {
namespace behaviors {

WakerBehavior::WakerBehavior(unsigned long timeout_ms) : timeoutPeriod(timeout_ms), waken(false) {}

WakerBehavior::WakerBehavior(tamaf::core::Agent* agent, unsigned long timeout_ms) : timeoutPeriod(timeout_ms), waken(false) {
    this->myAgent = agent;
}

void WakerBehavior::OnStart() {
    wakeupTime = millis() + timeoutPeriod;
    waken = false;
}

void WakerBehavior::Action() {
    if (millis() >= wakeupTime && !waken) {
        OnWake();
        waken = true;
    }
}

bool WakerBehavior::Done() {
    return waken;
}

} // namespace behaviors
} // namespace tamaf
